### PR TITLE
accessPolicy: Implement Put, Get, Delete access policy.

### PR DESCRIPTION
**Pull request by harshavardhana**
Supporting - http://docs.aws.amazon.com/AmazonS3/latest/dev/access-policy-language-overview.html

Currently supports following actions.

```
   "*":                             true,
   "s3:*":                          true,
   "s3:GetObject":                  true,
   "s3:ListBucket":                 true,
   "s3:PutObject":                  true,
   "s3:CreateBucket":               true,
   "s3:GetBucketLocation":          true,
   "s3:DeleteBucket":               true,
   "s3:DeleteObject":               true,
   "s3:AbortMultipartUpload":       true,
   "s3:ListBucketMultipartUploads": true,
   "s3:ListMultipartUploadParts":   true,
```

Following conditions for "StringEquals" and "StringNotEquals"

```
   "s3:prefix", "s3:max-keys"
```

### CODE REVIEW

**Comment by harshavardhana** [comment]
Need to implement 

- longest prefix order
- validate incoming json based on resources based on action map. 
-

**Comment by harshavardhana** [comment]
- By default all resources inherit Deny effect (anonymous requests).
- Deny takes precedence over Allow effect for all statements.
- Allow effect is explicitly required to make a resource public.

**Comment by harshavardhana** [comment]
Move buckets/policies/ --> `buckets/<bucketname>/access-policy.json`

File: bucket-handlers.go [hunk 1]
```diff
@@ -356,65 +385,51 @@ func (api storageAPI) PutBucketHandler(w http.ResponseWriter, r *http.Request) {
...
+				errorIf(err.Trace(r.URL.String()), "Presigned signature verification failed.", nil)
+				writeErrorResponse(w, r, SignatureDoesNotMatch, r.URL.Path)
+				return
+			}
+			if !ok {
+				writeErrorResponse(w, r, SignatureDoesNotMatch, r.URL.Path)
+				return
+			}
+		} else if isRequestSignatureV4(r) {
+			// Verify signature for the incoming body if any.
```
Comment (abperiasamy) [inline bucket-handlers.go]: In many places, "else if" cases have no default case. Such ignored errors are hard to trace.

File: auth-handler.go [hunk 2]
```diff
@@ -61,41 +61,76 @@ func isRequestPresignedSignatureV4(r *http.Request) bool {
...
+type authType int
+
+// List of all supported auth types.
+const (
+	unknownRequest authType = iota
+	anonymousRequest
+	presignedRequest
+	postPolicyRequest
+	signedRequest
+	jwtRequest
```
Comment (abperiasamy) [inline auth-handler.go]: authTypeUnknown
authTypeAnynomous
..

File: bucket-handlers.go [hunk 3]
```diff
@@ -26,14 +26,45 @@ import (
...
 	mux "github.com/gorilla/mux"
 	"github.com/minio/minio/pkg/crypto/sha256"
 	"github.com/minio/minio/pkg/fs"
 	"github.com/minio/minio/pkg/probe"
+	"github.com/minio/minio/pkg/s3/access"
 	"github.com/minio/minio/pkg/s3/signature4"
 )
 
+// http://docs.aws.amazon.com/AmazonS3/latest/dev/mpuAndPermissions.html
+func checkAccess(action string, bucket string, u *url.URL) (isAllowed bool, s3Error int) {
```
Comment (abperiasamy) [inline bucket-handlers.go]: Rename to enforceBucketPolicy

File: bucket-policy-handlers.go [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+	"github.com/minio/minio/pkg/fs"
+	"github.com/minio/minio/pkg/probe"
+	"github.com/minio/minio/pkg/s3/access"
+)
+
+// maximum supported access policy size.
+const maxAccessPolicySize = 20 * 1024 * 1024 // 20KiB.
+
+// isActionValid - verify if action is valid.
+func isActionValid(action string, u *url.URL, statements []accesspolicy.Statement) bool {
```
Comment (abperiasamy) [inline bucket-policy-handlers.go]: checkBucketPolicyAction - Verify if this "action" is allowed in the bucket policy for corresponding resource.

**Comment by abperiasamy** [comment]
BucketPolicyEvalStatements -> BucketPolicyEvalStatement
-> BucketPolicyMatchStatement(Action, Resource, Conditions, Statement) returns Action

**Comment by abperiasamy** [comment]
In the next patch, bucketpolicy pkg should merge to main. Also S3 error codes should be prefixed with ErrS3.

**Pull request merged**